File: apps/backend/src/services/payment-events.ts
import { createHash } from "node:crypto"
import { db, paymentRecords, processedPaymentEvents } from "@yomi/db"
import { and, eq } from "drizzle-orm"

export function payloadHash(payload: string): string {
  return createHash("sha256").update(payload).digest("hex")
}

export async function recordPaymentEvent(input: {
  provider: string
  eventId: string
  eventType: string
  payloadHash: string
}): Promise<{ duplicate: boolean }> {
  const inserted = await db
    .insert(processedPaymentEvents)
    .values(input)
    .onConflictDoNothing()
    .returning({ id: processedPaymentEvents.id })

  return { duplicate: inserted.length === 0 }
}

export async function upsertPaymentRecord(input: {
  userId: string
  provider: string
  kind: "subscription" | "credit_pack"
  productKey: string
  providerCustomerId?: string | null
  providerOrderId?: string | null
  providerPaymentId?: string | null
  providerSubscriptionId?: string | null
  amountCents: number
  currency: string
  status: string
  metadata?: Record<string, unknown>
}): Promise<string | null> {
  if (input.providerOrderId) {
    const [existing] = await db
      .select({ id: paymentRecords.id })
      .from(paymentRecords)
      .where(
        and(
          eq(paymentRecords.provider, input.provider),
          eq(paymentRecords.providerOrderId, input.providerOrderId),
        ),
      )
      .limit(1)

    if (existing) {
      await db
        .update(paymentRecords)
        .set({
          status: input.status,
          providerPaymentId: input.providerPaymentId ?? null,
          providerCustomerId: input.providerCustomerId ?? null,
          providerSubscriptionId: input.providerSubscriptionId ?? null,
          amountCents: input.amountCents,
          currency: input.currency,
          metadata: input.metadata ?? null,
          updatedAt: new Date(),
        })
        .where(eq(paymentRecords.id, existing.id))
      return existing.id
    }
  }

  // No existing record found by providerOrderId — insert fresh
  const inserted = await db
    .insert(paymentRecords)
    .values({
      userId: input.userId,
      provider: input.provider,
      kind: input.kind,
      productKey: input.productKey,
      providerCustomerId: input.providerCustomerId ?? null,
      providerOrderId: input.providerOrderId ?? null,
      providerPaymentId: input.providerPaymentId ?? null,
      providerSubscriptionId: input.providerSubscriptionId ?? null,
      amountCents: input.amountCents,
      currency: input.currency,
      status: input.status,
      metadata: input.metadata ?? null,
    })
    .onConflictDoNothing()
    .returning({ id: paymentRecords.id })

  return inserted[0]?.id ?? null
}

File: apps/api/src/yomi/services/payment_events.py
"""Payment event dedup + payment record upserts.

Port of apps/api/src/services/payment-events.ts.
"""

from __future__ import annotations

import hashlib
from datetime import UTC, datetime

from sqlalchemy import select, update
from sqlalchemy.dialects.postgresql import insert as pg_insert
from sqlalchemy.ext.asyncio import AsyncSession

from yomi.db.models_app import PaymentRecord, ProcessedPaymentEvent

CreditMetadata = dict[str, object] | None


def payload_hash(payload: str) -> str:
    return hashlib.sha256(payload.encode()).hexdigest()


async def record_payment_event(
    session: AsyncSession,
    *,
    provider: str,
    event_id: str,
    event_type: str,
    payload_hash_value: str,
) -> dict[str, bool]:
    inserted = (
        await session.execute(
            pg_insert(ProcessedPaymentEvent)
            .values(
                provider=provider,
                event_id=event_id,
                event_type=event_type,
                payload_hash=payload_hash_value,
            )
            .on_conflict_do_nothing()
            .returning(ProcessedPaymentEvent.id)
        )
    ).scalar_one_or_none()
    return {"duplicate": inserted is None}


async def upsert_payment_record(
    session: AsyncSession,
    *,
    user_id: str,
    provider: str,
    kind: str,
    product_key: str,
    provider_customer_id: str | None = None,
    provider_order_id: str | None = None,
    provider_payment_id: str | None = None,
    provider_subscription_id: str | None = None,
    amount_cents: int,
    currency: str,
    status: str,
    metadata: CreditMetadata = None,
) -> str | None:
    if provider_order_id:
        existing = (
            await session.execute(
                select(PaymentRecord.id).where(
                    PaymentRecord.provider == provider,
                    PaymentRecord.provider_order_id == provider_order_id,
                ).limit(1)
            )
        ).scalar_one_or_none()

        if existing is not None:
            await session.execute(
                update(PaymentRecord)
                .where(PaymentRecord.id == existing)
                .values(
                    status=status,
                    provider_payment_id=provider_payment_id,
                    provider_customer_id=provider_customer_id,
                    provider_subscription_id=provider_subscription_id,
                    amount_cents=amount_cents,
                    currency=currency,
                    metadata_=metadata,
                    updated_at=datetime.now(UTC),
                )
            )
            return str(existing)

    inserted = (
        await session.execute(
            pg_insert(PaymentRecord)
            .values(
                user_id=user_id,
                provider=provider,
                kind=kind,
                product_key=product_key,
                provider_customer_id=provider_customer_id,
                provider_order_id=provider_order_id,
                provider_payment_id=provider_payment_id,
                provider_subscription_id=provider_subscription_id,
                amount_cents=amount_cents,
                currency=currency,
                status=status,
                metadata=metadata,
            )
            .on_conflict_do_nothing()
            .returning(PaymentRecord.id)
        )
    ).scalar_one_or_none()
    return str(inserted) if inserted is not None else None